docs/design/vsock-design.md

```diff
@@ -0,0 +1,208 @@
+
+When the virtio-vsock device model in Firecracker detects a connection
+request coming from the guest (a VIRTIO_VSOCK_OP_REQUEST packet), it will try
+to forward the connection to an AF_UNIX socket listening on the host, at
+`/path/to/vsock_{CID}:{PORT}`, where `{PORT}` is the destination port, as
+specified in the connection request packet. If no such socket exists, or
+no one is listening on it, a connection cannot be established, and
+a VIRTIO_VSOCK_OP_RST packet will be sent back to the guest.
+
+From the user perspective, these would be the steps taken to establish a
+communication channel:
+
+1. Host: At VM configuration time, add a virtio-vsock device with `CID` and
+   `PATH`;
+2. Host: create and listen on an AF_UNIX socket at `{PATH}_{CID}:{PORT}`;
+3. Guest: create an AF_VSOCK socket and issue a `connect()` call to `HOST_CID`
+   and `PORT`;
+4. Host: `accept()` the new connection.
+
+The channel is established between the sockets obtained at steps 4 (host)
+and 3 (guest).
+
+### Host-Initiated Connections
+
+Firecracker will be listening on an AF_UNIX socket, at `{PATH}_{CID}`. When
+the host needs to initiate a connection, it will connect to that Unix socket,
+then send a special *connect* data packet on it, specifying the destination
+AF_VSOCK port. Following that, the same connection will be forwarded by
+Firecracker to the guest software listening on that port, thus establishing
+the requested channel. If no one is listening, Firecracker will terminate the
+host connection.
+
+1. Host: At VM configuration time, add a virtio-vsock device with `CID` and
+   `PATH`;
+2. Guest: create an AF_VSOCK socket and `listen()` on `PORT`;
+3. Host: `connect()` to AF_UNIX at `{PATH}_{CID}`;
+4. Host: `send()` *connect {PORT}* data packet on that connection;
+5. Guest: `accept()` the new connection.
+
+The channel is established between the sockets obtained at steps 3 (host)
+and 5 (guest).
+
+# Other Ideas
+
+Multiple ideas have been considered and evaluated, before ariving at the
+proposed solution. These ideas are listed below, together with a brief
+description of their strong and weak points.
+
+## Ethernet Host To Ethernet Guest
+
+Using traditional network interfaces to communicate between guest and host
+would require no implementation effort, but it also wouldn't be compliant
+with our goal of providing a lightweight solution. Additional `netfilter`
+rules would have to be set up on the host, for each microVM, in order to
+enforce guest isolation. When considering our density goals, that means
+the host would have to handle thousands of `netfilter` rules - a process
+that puts too much strain on the host resources.
+
+## Single Guest AF_VSOCK To Host AF_UNIX
+
+A single-channel mechanism, that would offload multiplexing duties to the
+user, similar to virtio-console. This would require the smallest
+implementation effort, but would drastically increase the integration effort
+needed from the Firecracker users.
```

Comment:
True, not having multiple ports being supported would be an additional burden to handle from the user guest code.